ext/mbstring/mbstring.c

```diff
@@ -2588,22 +2588,29 @@ PHP_FUNCTION(mb_convert_case)
 /* {{{ Returns a upper cased version of source_string */
 PHP_FUNCTION(mb_strtoupper)
 {
+	zend_string *str;
 	zend_string *from_encoding = NULL;
-	char *str;
-	size_t str_len, ret_len;
+	const mbfl_encoding *enc;
+	char *newstr;
+	size_t ret_len;
 
 	ZEND_PARSE_PARAMETERS_START(1, 2)
-		Z_PARAM_STRING(str, str_len)
+		Z_PARAM_STR(str)
 		Z_PARAM_OPTIONAL
 		Z_PARAM_STR_OR_NULL(from_encoding)
 	ZEND_PARSE_PARAMETERS_END();
 
-	const mbfl_encoding *enc = php_mb_get_encoding(from_encoding, 2);
+	enc = php_mb_get_encoding(from_encoding, 2);
 	if (!enc) {
 		RETURN_THROWS();
 	}
 
-	char *newstr = mbstring_convert_case(PHP_UNICODE_CASE_UPPER, str, str_len, &ret_len, enc);
+	// optimize performance for UTF-8 encoding and input string consisting of lower/7-bit ASCII characters only
```

Comment:
Yeah, but that deoptimizes the performance for non ASCII strings. I think we need to measure this.

Reply:
The performance drop for non-ASCII strings is around 5%. Which is highly negligible, in short, one ASCII only string conversion gains this drop back for the next 20 non-ASCII conversions. 😅

Reply:
Okay, but how often would there be pure ASCII strings. In the worst case never.

Reply:
often :), consider this function for example for lowecasing a php class name to build a case-insensitive index (for short strings, there is "worst case gain" of 250%)